File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
using namespace std;

int m, n, k;
int grid[51][51];
int visited[51][51];


void initArr()
{
	for(int i=0; i<51;i++)
	{
		for(int j=0; j<51;j++)
		{
			grid[j][i]=0;
			visited[j][i]=0;
		}		
	}
}

void dfs(int x, int y)
{
	int dx[] = {0,0,-1,1};
	int dy[] = {1,-1 ,0,0};
			
	for(int i =0; i< 4;i++)
	{
		int nx = x + dx[i];
		int ny = y + dy[i];
		
		if(nx < 0 || nx >= m || ny < 0 || ny >= n) continue;
		if(visited[ny][nx] != 0) continue;
		if(grid[ny][nx] != 1) continue;
		
		visited[ny][nx] = 1;
		dfs(nx, ny);
	}
}

int main()
{
	int t;
	cin >> t;
	
	for(int i = 0; i < t; i++)
	{
		initArr();			
		
		cin >> m >> n >> k;				
		
		for(int j = 0; j < k; j++)
		{
			int x, y;
			cin >> x >> y;
			grid[y][x] = 1;
		}
				
		int count = 0;
		for(int j=0;j<n;j++)
		{
			for(int l=0;l<m;l++)
			{
				if(visited[j][l] != 0) continue;
				if(grid[j][l] != 1) continue;
				
				visited[j][l] = 1;
				dfs(l, j);
				count++;
			}
		}
		
		cout << count << '\n';
	}	

	return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.cs
using System;

public class Sol
{
    static int[] dx = new int[]{0,0,-1,1};
    static int[] dy = new int[]{1,-1,0,0};
    
    static void Dfs(int x, int y, int[,] grid, int m, int n)
    {
        grid[y,x] = 0;
        
        for(int i =0; i< 4;i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(nx < 0 || ny < 0 || nx >= m || ny >= n) continue;
            if(grid[ny,nx] != 1) continue;
            
            Dfs(nx, ny, grid,m,n);
        }
    }
    
    static public void Main(string[] args)
    {
        int t = int.Parse(Console.ReadLine());
        for(int tmp = 0; tmp < t; tmp++)
        {
            int count = 0;
            string[] arr = Console.ReadLine().Split();
            
            int n,m,k;
            m = int.Parse(arr[0]);
            n = int.Parse(arr[1]);
            k = int.Parse(arr[2]);
            
            int[,] grid = new int[n,m];
            
            for(int i = 0; i < k; i++)
            {
                string[] strs = Console.ReadLine().Split();
                grid[int.Parse(strs[1]),int.Parse(strs[0])] = 1; // 1은 배추
            }
            
            for(int i =0; i<n;i++)
            {
                for(int j =0;j<m;j++)
                {
                    if(grid[i,j]!=1)continue;
                    Dfs(j,i,grid,m,n);
                    count++;
                }
            }
        
            Console.WriteLine(count.ToString());
        }
    }
}